### PR TITLE
Adds functionality to diff BlameReports

### PR DESCRIPTION
**Pull request by vulder**
Adds a BlameReportDiff class that represents the diff between two
BlameReports and shows all interactions that changed.

### CODE REVIEW

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- varats/data/reports/blame_report.py  10
         

Clones added
============
- tests/data/test_blame_report.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5467441&bid=17919615)

File: varats/data/reports/blame_report.py [hunk 1]
```diff
@@ -105,6 +150,39 @@ def __str__(self) -> str:
         return str_representation
 
 
+def _calc_diff_between_func_entries(
```
Comment (boehmseb) [inline varats/data/reports/blame_report.py:154]: Maybe a matter of taste, but is there a reason why you keep this part of the diffing logic in the data-holding class while the rest is in the diffing-class?

File: varats/data/reports/blame_report.py [hunk 2]
```diff
@@ -105,6 +150,39 @@ def __str__(self) -> str:
...
+        prev_func_entry: BlameResultFunctionEntry) -> BlameResultFunctionEntry:
+    diff_interactions: tp.List[BlameInstInteractions] = []
+
+    base_interactions = sorted(base_func_entry.interactions)
+    prev_interactions = sorted(prev_func_entry.interactions)
+
+    for base_inter in base_interactions:
+        if base_inter in prev_interactions:
+            prev_inter_idx = prev_interactions.index(base_inter)
+            prev_inter = prev_interactions[prev_inter_idx]
```
Comment (boehmseb) [inline varats/data/reports/blame_report.py]:
```suggestion
            prev_inter = prev_interactions.pop(prev_inter_idx)
```
And remove the `del` below.

File: varats/data/reports/blame_report.py [hunk 3]
```diff
@@ -105,6 +150,39 @@ def __str__(self) -> str:
...
+            # create new blame inst interaction with the absolute differente
+            # between base and prev
+            difference = abs(base_inter.amount - prev_inter.amount)
+            if difference > 0:
+                diff_interactions.append(
+                    BlameInstInteractions(
+                        base_inter.base_commit,
+                        deepcopy(base_inter.interacting_commits), difference))
+            del prev_interactions[prev_inter_idx]
```
Comment (boehmseb) [inline varats/data/reports/blame_report.py]: Why use `del` here? I would suggest to use `pop()` above and remove this line.

File: varats/data/reports/blame_report.py [hunk 4]
```diff
@@ -189,6 +268,103 @@ def __str__(self) -> str:
         return str_representation
 
 
+class BlameReportDiff():
+    """
+    Diff class that contains all interactions that changed between to report
```
Comment (boehmseb) [inline varats/data/reports/blame_report.py]:
```suggestion
    Diff class that contains all interactions that changed between two report
```

**Review by boehmseb — CHANGES_REQUESTED** [review]
The diff logic itself looks fine to me. Just one typo and a design-related comment.

File: varats/data/reports/blame_report.py (hunk 1 above)
Comment (vulder) [reply varats/data/reports/blame_report.py:154]: As discussed we keep this for now

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- varats/data/reports/blame_report.py  10
         

Clones added
============
- tests/data/test_blame_report.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5467441&bid=17919615)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- varats/data/reports/blame_report.py  10
         

Clones added
============
- tests/data/test_blame_report.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5467441&bid=17919615)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- varats/data/reports/blame_report.py  10
         

Clones added
============
- tests/data/test_blame_report.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5467441&bid=17919615)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- varats/data/reports/blame_report.py  10
         

Clones added
============
- tests/data/test_blame_report.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5467441&bid=17919615)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- varats/data/reports/blame_report.py  10
         

Clones added
============
- tests/data/test_blame_report.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5467441&bid=17919615)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- varats/data/reports/blame_report.py  10
         

Clones added
============
- tests/data/test_blame_report.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5467441&bid=17919615)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- varats/data/reports/blame_report.py  10
         

Clones added
============
- tests/data/test_blame_report.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5467441&bid=17919615)

**Review by boehmseb — APPROVED** [review]

File: .travis.yml [hunk 5]
```diff
@@ -23,7 +23,7 @@ install:
 
 script:
   - mkdir -p paper_configs/test/
-  - git clone https://git.savannah.gnu.org/git/gzip.git gzip
+  # - git clone https://git.savannah.gnu.org/git/gzip.git gzip
```
Comment (vulder) [inline .travis.yml]:
```suggestion
```

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- varats/data/reports/blame_report.py  10
         

Clones added
============
- tests/data/test_blame_report.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5467441&bid=17919615)

**Pull request merged**